### PR TITLE
support lease Attach

### PR DESCRIPTION
**Pull request by xiang90**
Now we can attach keys to leases. And revoking the lease removes all
the attached keys of that lease.

### CODE REVIEW

**Comment by xiang90** [comment]
@heyitsanthony I feel the SetDeletableRange interface is ugly. But I do not have a better way right now. Please give it a look.

**Comment by heyitsanthony** [comment]
An alternative (not necessarily better): the Lessor interface has a `DeleteChan() <-chan Range` and a `UpdateChan() chan<- RevID`. Deletes are posted to `DeleteChan()` by the lessor and the results are posted to `UpdateChan()` by the backend. This avoids having to do any Set* stuff, at least.

Otherwise, maybe call it `RangeDeleter` instead of `DeleteableRange` to make it clear that it's not a range but something that acts on ranges.

**Comment by xiang90** [comment]
@heyitsanthony OK. I will try that idea out.

**Comment by xiang90** [comment]
@heyitsanthony All fixed. PTAL.

**Comment by xiang90** [comment]
@heyitsanthony Fixing the test failure now.

File: storage/kvstore.go [hunk 1]
```diff
@@ -397,6 +417,17 @@ func (s *store) put(key, value []byte, lease lease.LeaseID) {
 	s.tx.UnsafePut(keyBucketName, ibytes, d)
 	s.kvindex.Put(key, revision{main: rev, sub: s.currentRev.sub})
 	s.currentRev.sub += 1
+
+	if s.le != nil && leaseID != lease.NoLease {
+		// TODO: validate the existence of lease before call Attach.
```
Comment (heyitsanthony) [inline storage/kvstore.go]: will this just ignore the lease if there's a leaseID but no lessor?

File: storage/kvstore.go [hunk 2]
```diff
@@ -283,9 +292,20 @@ func (s *store) restore() error {
 		// restore index
 		switch {
 		case isTombstone(key):
+			// TODO: De-attach keys from lease if necessary
 			s.kvindex.Tombstone(kv.Key, rev)
 		default:
 			s.kvindex.Restore(kv.Key, revision{kv.CreateRevision, 0}, rev, kv.Version)
+			if s.le != nil && lease.LeaseID(kv.Lease) != lease.NoLease {
```
Comment (heyitsanthony) [inline storage/kvstore.go]: what happens if there's a lease but no lessor? is it ok to ignore the lease?

File: storage/kv.go [hunk 3]
```diff
@@ -68,6 +68,7 @@ type KV interface {
 
 	// Restore restores the KV store from a backend.
 	Restore(b backend.Backend) error
+
```
Comment (heyitsanthony) [inline storage/kv.go]: extra line

**Comment by heyitsanthony** [comment]
the `leaseid != nolease && le == nil` case might need more consideration

File: storage/kvstore.go (hunk 1 above)
Comment (xiang90) [inline storage/kvstore.go]: will panic on s.le but leaseID exists.

File: storage/kvstore.go (hunk 2 above)
Comment (xiang90) [inline storage/kvstore.go]: should panic.

File: storage/kv.go (hunk 3 above)
Comment (xiang90) [inline storage/kv.go]: will remove

File: storage/kvstore.go [hunk 4]
```diff
@@ -283,9 +292,23 @@ func (s *store) restore() error {
 		// restore index
 		switch {
 		case isTombstone(key):
+			// TODO: De-attach keys from lease if necessary
 			s.kvindex.Tombstone(kv.Key, rev)
 		default:
 			s.kvindex.Restore(kv.Key, revision{kv.CreateRevision, 0}, rev, kv.Version)
+			if lease.LeaseID(kv.Lease) != lease.NoLease {
+				if s.le == nil {
```
Comment (xiang90) [inline storage/kvstore.go]: @heyitsanthony added panic on this case

**Comment by xiang90** [comment]
@heyitsanthony Addressed the issues. PTAL.

**Comment by heyitsanthony** [comment]
ok overall, just fix ci issues

**Comment by xiang90** [comment]
@heyitsanthony Sure. Will fix.

**Comment by xiang90** [comment]
@heyitsanthony CI turned green. Merging.

**Pull request merged**